fix(component): `@layout` accept template factory

buschtoens opened the pull request:
https://github.com/emberjs/ember.js/pull/18096

This PR changed how templates get transpiled under the hood.
The export from templates are now factory functions which expect to be passed the `owner`.

rwjblue commented:
CI is failing due to using Node 6:

```
@ember-data/-build-infra/node_modules/broccoli-rollup/dist/index.js:31
    async build() {
          ^^^^^
```

buschtoens commented:
Should we pin the Ember Data dependency for now and do a major release later that drops Node 6?

buschtoens commented on packages/component/addon/index.js:
```diff
@@ -201,7 +201,9 @@ export const layout = decoratorWithRequiredParams((target, params) => {
-    (() => typeof template === 'object' && typeof template.indexOf === 'undefined')()
+    (() =>
+      typeof template === 'function' ||
+      (typeof template === 'object' && typeof template.indexOf === 'undefined'))()
```
btw is there any reason this is an IIFE? If not, I'd remove that in this PR as well.

rwjblue replied:
No not particularly. I'd only use an IIFE in order to simplify the boolean logic, e.g.:

```js
assert(
  'whatever',
  (() => {
    if (condition) { return true; }
    if (otherThing) { return true; }
    return false;
  })()
```

It would probably make this specific case easier to think about, but isn't a giant deal.

buschtoens commented:
I am sure that the [`typescript@next` error](https://travis-ci.org/ember-decorators/ember-decorators/jobs/554125248#L550-L552) is caused by https://github.com/microsoft/TypeScript/pull/30790.

```
Error: Errors in typescript@next for external dependencies:
../../../node_modules/@types/node/index.d.ts(187,11): error TS2300: Duplicate identifier 'IteratorResult'.
../../../node_modules/dtslint/typescript-installs/next/node_modules/typescript/lib/lib.es2015.iterable.d.ts(41,6): error TS2300: Duplicate identifier 'IteratorResult'.
```

I reported it here: https://github.com/microsoft/TypeScript/pull/30790#issuecomment-508411128

buschtoens commented:
Oh come on FFS, how hard can it be to get a freaking one-line change merged? 😂

`ember-cli-addon-docs` and all the dependencies it brings in are just causing _sooo_ many issues. I'll open another PR to get `master` clean first, when I can find some time.

buschtoens commented:
Alright, I discovered another issue with https://github.com/emberjs/ember.js/pull/18096 in combination with `@layout` in production builds. It's a reprise of #327.

In development builds or builds that only target browsers that support arrow functions, the returned `TemplateFactory` and thus first argument passed to `@layout` is an arrow function, e.g. `owner => { ... }`.

Importantly, arrow functions have no `prototype` key, while regular functions (`function(owner) { ... }`) do.

https://github.com/ember-decorators/ember-decorators/blob/9f6af93075f1c2c216759e4e763ab487d382958b/packages/utils/addon/-private/class-field-descriptor.js#L1-L10

This causes the above code to report a false positive in production builds an lets the `@layout` decorator fail with:

> Error: Assertion Failed: The `@layout` decorator requires parameters

Since [`decoratorWithRequiredParams`](https://github.com/ember-decorators/ember-decorators/blob/9f6af93075f1c2c216759e4e763ab487d382958b/packages/utils/addon/decorator.js#L50-L59) thinks the decorator is invoked like:

```ts
@layout
class {}

// instead of
@layout(template)
class {}
```

I suggest we stop using `decoratorWithRequiredParams` for `@layout`. Unfortunately we use a check here then, but I don't see any other way to fix this.

rwjblue commented:
Ya, that seems fine to me...

josemarluedke commented:
@buschtoens @rwjblue  Is there anything that I can help with to move this forward? I'm currently blocked on upgrading Ember Canary because of the issue this solves.

rwjblue commented:
AFAIK the blocker is just implementing the change @buschtoens suggested

buschtoens commented:
I fixed the `decoratorWithRequiredParams` usage and removed the IIFE, but I expect CI to still fail spectacularly, because of all the other unrelated issues 🙈

rwjblue commented:
@pzuraq - Thoughts on the path forward here/

pzuraq commented:
I think the solution looks good, I'm not actually sure what the test failures on Canary are, it looks like they're coming from pretty deep inside the VM O_o 

Also unsure what's going on with production tests. I can try to take a look sometime next week, going to be busy for the remainder of this week. I also think we could make canary tests optional at this point, since we don't expect much to be changing with regards to the object model in the near future. Ideally we wouldn't, but we could to get unblocked.

buschtoens commented:
The canary failure is caused by https://github.com/emberjs/ember-test-helpers/pull/677

rwjblue commented:
Ya, could be but I thought I fixed those issues in ember so it didn't require changes in addons (e.g. didn't need to upgrade `@ember/test-helpers`).

josemarluedke commented:
Failures are agast "canary" and "production", could it be that when the tests ran, the build didn't pick up the changes that @rwjblue is referring as fixed? 

Error: `Cannot read property 'lookup' of undefined`

The error does seem to be related, maybe some can restart the tests to see if it passes now? 

Looking forward to having this fixed.

rwjblue commented:
The production build failure was not related to the lookup failure (it was `Error: Browser failed to connect within 30s. testem.js not loaded?`).

Apparently my fix in ember-source only fixed `setupRenderingTest` style tests, not `moduleForComponent`. I haven't had a chance to dig into that yet though. It is blocking a couple legit bug fixes over in @ember/test-helpers though.

rwjblue commented:
A plausible path forward here would be to run the codemod to migrate to setupRenderingTest to work around the bug in moduleForComponent.

rwjblue commented:
OK, now ready for a rebase...

rwjblue commented:
Rebased and merged in #454
